File: test/layout/tools/test_layout_auto.py
#!/usr/bin/env python3
"""
Radiant Layout Engine Automated Testing

This script provides automated layout validation for the Radiant layout engine
by running layout tests via CLI and comparing results against browser references.
"""

import os
import sys
import json
import subprocess
import time
import argparse
from pathlib import Path
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass, asdict


@dataclass
class TestResult:
    """Result of a single layout test"""
    test_name: str
    passed: bool
    error_message: Optional[str] = None
    execution_time: float = 0.0
    matched_elements: int = 0
    total_elements: int = 0
    max_difference: float = 0.0
    differences: List[Dict] = None
    
    def __post_init__(self):
        if self.differences is None:
            self.differences = []


@dataclass
class TestSummary:
    """Summary of all test results"""
    total_tests: int = 0
    passed_tests: int = 0
    failed_tests: int = 0
    error_tests: int = 0
    total_time: float = 0.0
    results: List[TestResult] = None
    
    def __post_init__(self):
        if self.results is None:
            self.results = []
    
    @property
    def pass_rate(self) -> float:
        return (self.passed_tests / self.total_tests * 100) if self.total_tests > 0 else 0.0


class RadiantLayoutTester:
    """Main class for running Radiant layout tests"""
    
    def __init__(self, radiant_exe: str = "./radiant.exe", tolerance: float = 2.0):
        self.radiant_exe = radiant_exe
        self.tolerance = tolerance
        self.test_dir = Path("test/layout")
        self.data_dir = self.test_dir / "data"
        self.reference_dir = self.test_dir / "reference"
        self.reports_dir = self.test_dir / "reports"
        
        # Ensure directories exist
        self.reports_dir.mkdir(parents=True, exist_ok=True)
        
    def discover_tests(self, category: Optional[str] = None) -> List[Path]:
        """Discover available test HTML files"""
        test_files = []
        
        if category:
            categories = [category]
        else:
            categories = ["basic", "intermediate", "advanced"]
        
        for cat in categories:
            cat_dir = self.data_dir / cat
            if cat_dir.exists():
                html_files = list(cat_dir.glob("*.html"))
                test_files.extend(html_files)
        
        return sorted(test_files)
    
    def run_single_test(self, html_file: Path) -> TestResult:
        """Run a single layout test"""
        test_name = html_file.stem
        category = html_file.parent.name
        
        print(f"  🧪 Testing: {test_name}")
        
        start_time = time.time()
        
        try:
            # Run Radiant layout command
            result = subprocess.run(
                [self.radiant_exe, "layout", str(html_file)],
                capture_output=True,
                text=True,
                timeout=30,  # 30 second timeout
                cwd=Path.cwd()
            )
            
            execution_time = time.time() - start_time
            
            if result.returncode != 0:
                return TestResult(
                    test_name=test_name,
                    passed=False,
                    error_message=f"Radiant execution failed (exit code {result.returncode}): {result.stderr}",
                    execution_time=execution_time
                )
            
            # Check if output files were generated
            view_tree_json = Path("view_tree.json")
            view_tree_txt = Path("view_tree.txt")
            
            if not view_tree_json.exists():
                return TestResult(
                    test_name=test_name,
                    passed=False,
                    error_message="view_tree.json not generated",
                    execution_time=execution_time
                )
            
            # Load Radiant output
            try:
                with open(view_tree_json, 'r') as f:
                    radiant_data = json.load(f)
            except json.JSONDecodeError as e:
                return TestResult(
                    test_name=test_name,
                    passed=False,
                    error_message=f"Invalid JSON in view_tree.json: {e}",
                    execution_time=execution_time
                )
            
            # Load browser reference (if available)
            reference_file = self.reference_dir / category / f"{test_name}.json"
            if reference_file.exists():
                try:
                    with open(reference_file, 'r') as f:
                        reference_data = json.load(f)
                    
                    # Compare layouts
                    comparison = self.compare_layouts(radiant_data, reference_data)
                    
                    test_result = TestResult(
                        test_name=test_name,
                        passed=comparison["max_difference"] <= self.tolerance,
                        execution_time=execution_time,
                        matched_elements=comparison["matched_elements"],
                        total_elements=comparison["total_elements"],
                        max_difference=comparison["max_difference"],
                        differences=comparison["differences"]
                    )
                    
                except json.JSONDecodeError as e:
                    test_result = TestResult(
                        test_name=test_name,
                        passed=False,
                        error_message=f"Invalid reference JSON: {e}",
                        execution_time=execution_time
                    )
            else:
                # No reference data - just validate that Radiant produced output
                elements_count = self.count_elements(radiant_data.get("layout_tree", {}))
                test_result = TestResult(
                    test_name=test_name,
                    passed=True,
                    execution_time=execution_time,
                    matched_elements=elements_count,
                    total_elements=elements_count,
                    max_difference=0.0
                )
                print(f"    ⚠️  No reference data - validated Radiant execution only")
            
            # Cleanup output files
            if view_tree_json.exists():
                view_tree_json.unlink()
            if view_tree_txt.exists():
                view_tree_txt.unlink()
            
            # Print result
            if test_result.passed:
                print(f"    ✅ PASS ({test_result.matched_elements}/{test_result.total_elements} elements)")
            else:
                print(f"    ❌ FAIL ({test_result.matched_elements}/{test_result.total_elements} elements)")
                if test_result.max_difference > 0:
                    print(f"       Max difference: {test_result.max_difference:.2f}px")
            
            return test_result
            
        except subprocess.TimeoutExpired:
            return TestResult(
                test_name=test_name,
                passed=False,
                error_message="Test timed out after 30 seconds",
                execution_time=time.time() - start_time
            )
        except Exception as e:
            return TestResult(
                test_name=test_name,
                passed=False,
                error_message=f"Unexpected error: {str(e)}",
                execution_time=time.time() - start_time
            )
    
    def count_elements(self, layout_tree: Dict) -> int:
        """Count elements in layout tree"""
        if not layout_tree:
            return 0
        
        count = 1  # Count this element
        children = layout_tree.get("children", [])
        for child in children:
            count += self.count_elements(child)
        
        return count
    
    def compare_layouts(self, radiant_data: Dict, reference_data: Dict) -> Dict:
        """Compare Radiant layout with browser reference"""
        differences = []
        max_difference = 0.0
        matched_elements = 0
        total_elements = 0
        
        # Extract elements from both layouts
        radiant_elements = self.extract_elements(radiant_data.get("layout_tree", {}))
        reference_elements = reference_data.get("layout_data", {})
        
        # Compare each element
        for selector, radiant_element in radiant_elements.items():
            total_elements += 1
            
            if selector in reference_elements:
                ref_element = reference_elements[selector]
                element_diff = self.compare_element(radiant_element, ref_element)
                
                if element_diff["max_diff"] <= self.tolerance:
                    matched_elements += 1
                else:
                    differences.append({
                        "selector": selector,
                        "differences": element_diff["differences"],
                        "max_diff": element_diff["max_diff"]
                    })
                
                max_difference = max(max_difference, element_diff["max_diff"])
        
        return {
            "matched_elements": matched_elements,
            "total_elements": total_elements,
            "max_difference": max_difference,
            "differences": differences
        }
    
    def extract_elements(self, layout_tree: Dict, path: str = "", elements: Dict = None) -> Dict:
        """Extract elements from layout tree with CSS-like selectors"""
        if elements is None:
            elements = {}
        
        if not layout_tree:
            return elements
        
        # Generate selector for this element
        selector = path or "root"
        
        elements[selector] = {
            "layout": layout_tree.get("layout", {}),
            "css_properties": layout_tree.get("css_properties", {}),
            "type": layout_tree.get("type", "unknown"),
            "tag": layout_tree.get("tag", "unknown")
        }
        
        # Process children
        children = layout_tree.get("children", [])
        for index, child in enumerate(children):
            child_path = f"{selector} > :nth-child({index + 1})"
            self.extract_elements(child, child_path, elements)
        
        return elements
    
    def compare_element(self, radiant_element: Dict, reference_element: Dict) -> Dict:
        """Compare a single element between Radiant and reference"""
        differences = []
        max_diff = 0.0
        
        # Compare layout properties
        layout_props = ["x", "y", "width", "height"]
        radiant_layout = radiant_element.get("layout", {})
        ref_layout = reference_element.get("layout", {})
        
        for prop in layout_props:
            radiant_value = radiant_layout.get(prop, 0)
            ref_value = ref_layout.get(prop, 0)
            diff = abs(radiant_value - ref_value)
            
            if diff > 0.1:  # Ignore tiny differences
                differences.append({
                    "property": prop,
                    "radiant": radiant_value,
                    "reference": ref_value,
                    "difference": diff
                })
                max_diff = max(max_diff, diff)
        
        return {"differences": differences, "max_diff": max_diff}
    
    def run_category(self, category: str) -> TestSummary:
        """Run all tests in a category"""
        print(f"📂 Processing {category} tests...")
        
        test_files = self.discover_tests(category)
        if not test_files:
            print(f"  ⚠️  No test files found in {category}")
            return TestSummary()
        
        summary = TestSummary()
        start_time = time.time()
        
        for test_file in test_files:
            result = self.run_single_test(test_file)
            summary.results.append(result)
            summary.total_tests += 1
            
            if result.passed:
                summary.passed_tests += 1
            elif result.error_message:
                summary.error_tests += 1
            else:
                summary.failed_tests += 1
        
        summary.total_time = time.time() - start_time
        
        print(f"  📊 Category results: {summary.passed_tests}/{summary.total_tests} passed ({summary.pass_rate:.1f}%)")
        print()
        
        return summary
    
    def run_all_tests(self) -> TestSummary:
        """Run all available tests"""
        print("🎨 Radiant Layout Validation Suite")
        print("===================================\n")
        
        # Check if Radiant executable exists
        if not Path(self.radiant_exe).exists():
            print(f"❌ Error: Radiant executable not found: {self.radiant_exe}")
            print("Please build Radiant first with: make build-radiant")
            sys.exit(1)
        
        categories = ["basic", "intermediate", "advanced"]
        overall_summary = TestSummary()
        
        for category in categories:
            if (self.data_dir / category).exists():
                category_summary = self.run_category(category)
                
                # Merge results
                overall_summary.total_tests += category_summary.total_tests
                overall_summary.passed_tests += category_summary.passed_tests
                overall_summary.failed_tests += category_summary.failed_tests
                overall_summary.error_tests += category_summary.error_tests
                overall_summary.total_time += category_summary.total_time
                overall_summary.results.extend(category_summary.results)
        
        self.print_summary(overall_summary)
        self.save_report(overall_summary)
        
        return overall_summary
    
    def print_summary(self, summary: TestSummary):
        """Print test results summary"""
        print("📊 Test Results Summary")
        print("=======================")
        print(f"Total tests: {summary.total_tests}")
        print(f"✅ Passed: {summary.passed_tests}")
        print(f"❌ Failed: {summary.failed_tests}")
        print(f"💥 Errors: {summary.error_tests}")
        print(f"⏱️  Total time: {summary.total_time:.2f}s")
        
        if summary.total_tests > 0:
            print(f"📈 Pass rate: {summary.pass_rate:.1f}%")
        
        # Show failed tests
        failed_results = [r for r in summary.results if not r.passed]
        if failed_results:
            print(f"\n❌ Failed Tests ({len(failed_results)}):")
            for result in failed_results[:10]:  # Show first 10 failures
                print(f"  • {result.test_name}: {result.error_message or 'Layout differences'}")
                if result.max_difference > 0:
                    print(f"    Max difference: {result.max_difference:.2f}px")
            
            if len(failed_results) > 10:
                print(f"  ... and {len(failed_results) - 10} more failures")
    
    def save_report(self, summary: TestSummary):
        """Save detailed test report"""
        timestamp = int(time.time())
        report_file = self.reports_dir / f"radiant_validation_{timestamp}.json"
        
        report_data = {
            "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
            "radiant_exe": self.radiant_exe,
            "tolerance": self.tolerance,
            "summary": asdict(summary)
        }
        
        with open(report_file, 'w') as f:
            json.dump(report_data, f, indent=2)
        
        print(f"📄 Detailed report saved: {report_file}")


def create_sample_test_files():
    """Create sample test HTML files for demonstration"""
    basic_dir = Path("test/layout/data/basic")
    basic_dir.mkdir(parents=True, exist_ok=True)
    
    # Basic flex test
    flex_test = """<!DOCTYPE html>
<html>
<head>
    <style>
        .container {
            display: flex;
            flex-direction: row;
            justify-content: space-between;
            width: 600px;
            height: 100px;
            background: #f0f0f0;
        }
        .item {
            width: 100px;
            height: 60px;
            background: #4CAF50;
            flex-shrink: 0;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="item"></div>
        <div class="item"></div>
        <div class="item"></div>
    </div>
</body>
</html>"""
    
    with open(basic_dir / "flex_001_row_space_between.html", 'w') as f:
        f.write(flex_test)
    
    # Basic block test
    block_test = """<!DOCTYPE html>
<html>
<head>
    <style>
        .container {
            width: 400px;
            height: 200px;
            padding: 20px;
            margin: 10px;
            background: #e3f2fd;
        }
        .block {
            width: 200px;
            height: 80px;
            margin: 10px 0;
            padding: 15px;
            background: #2196f3;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="block">Block 1</div>
        <div class="block">Block 2</div>
    </div>
</body>
</html>"""
    
    with open(basic_dir / "block_001_margin_padding.html", 'w') as f:
        f.write(block_test)
    
    print("✅ Sample test files created in test/layout/data/basic/")


def main():
    parser = argparse.ArgumentParser(description="Radiant Layout Engine Automated Testing")
    parser.add_argument("--category", choices=["basic", "intermediate", "advanced"],
                       help="Run tests for specific category only")
    parser.add_argument("--tolerance", type=float, default=2.0,
                       help="Tolerance in pixels for layout differences (default: 2.0)")
    parser.add_argument("--radiant-exe", default="./radiant.exe",
                       help="Path to Radiant executable (default: ./radiant.exe)")
    parser.add_argument("--create-samples", action="store_true",
                       help="Create sample test files")
    parser.add_argument("--verbose", "-v", action="store_true",
                       help="Enable verbose output")
    
    args = parser.parse_args()
    
    if args.create_samples:
        create_sample_test_files()
        return
    
    # Change to project root directory
    if Path("radiant").exists() and Path("test").exists():
        # We're already in the right directory
        pass
    elif Path("../radiant").exists():
        # We're in a subdirectory, go up one level
        os.chdir("..")
    elif Path("../../radiant").exists():
        # We're in test/layout/tools, go up three levels
        os.chdir("../../..")
    elif Path("../../../radiant").exists():
        # We're in a deeper subdirectory
        os.chdir("../../../..")
    else:
        print("❌ Error: Please run from the Lambda project root directory")
        print(f"Current directory: {Path.cwd()}")
        print("Expected to find 'radiant' and 'test' directories")
        sys.exit(1)
    
    tester = RadiantLayoutTester(
        radiant_exe=args.radiant_exe,
        tolerance=args.tolerance
    )
    
    if args.category:
        summary = tester.run_category(args.category)
    else:
        summary = tester.run_all_tests()
    
    # Exit with error code if tests failed
    if summary.failed_tests > 0 or summary.error_tests > 0:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: test/layout/tools/test_layout_auto.cpp
/**
 * GTest Integration for Radiant Layout Engine Testing
 * 
 * Provides Google Test integration for automated layout validation
 * against browser reference data.
 */

#include <gtest/gtest.h>
#include "layout_test_framework.hpp"
#include <memory>
#include <string>
#include <vector>
#include <map>
#include <filesystem>

using namespace LayoutTest;

class LayoutTestSuite : public ::testing::Test {
protected:
    void SetUp() override {
        testRunner_ = std::make_unique<TestRunner>();
        testRunner_->setVerbose(false); // Controlled by GTest verbosity
        
        // Initialize test directories
        testDataDir_ = "./data";
        testReferenceDir_ = "./reference";
        testReportsDir_ = "./reports";
        
        // Ensure directories exist
        std::filesystem::create_directories(testReportsDir_);
    }
    
    void TearDown() override {
        // Cleanup if needed
    }
    
    // Helper to create parametric test data
    static std::vector<TestCase> getAllTestCases() {
        TestRunner runner;
        return runner.discoverTests(".");
    }
    
    static std::vector<TestCase> getTestCasesByCategory(const std::string& category) {
        TestRunner runner;
        return runner.loadCategory(category);
    }
    
    std::unique_ptr<TestRunner> testRunner_;
    std::string testDataDir_;
    std::string testReferenceDir_;
    std::string testReportsDir_;
};

// Basic smoke tests
TEST_F(LayoutTestSuite, FrameworkInitialization) {
    ASSERT_NE(testRunner_, nullptr);
    
    // Test directory structure exists
    EXPECT_TRUE(std::filesystem::exists(testDataDir_)) 
        << "Test data directory should exist: " << testDataDir_;
    EXPECT_TRUE(std::filesystem::exists(testReferenceDir_)) 
        << "Reference data directory should exist: " << testReferenceDir_;
}

TEST_F(LayoutTestSuite, TestDiscovery) {
    auto allTests = testRunner_->discoverTests(".");
    EXPECT_GT(allTests.size(), 0) << "Should discover at least one test case";
    
    // Check that we have tests in each category
    auto basicTests = testRunner_->loadCategory("basic");
    auto intermediateTests = testRunner_->loadCategory("intermediate");
    auto advancedTests = testRunner_->loadCategory("advanced");
    
    EXPECT_GT(basicTests.size(), 0) << "Should have basic test cases";
    EXPECT_GT(intermediateTests.size(), 0) << "Should have intermediate test cases";
    EXPECT_GT(advancedTests.size(), 0) << "Should have advanced test cases";
}

// Parametric test fixture for running individual test cases
class LayoutTestCase : public LayoutTestSuite, 
                      public ::testing::WithParamInterface<TestCase> {
};

TEST_P(LayoutTestCase, ValidateLayout) {
    TestCase testCase = GetParam();
    
    SCOPED_TRACE("Testing: " + testCase.name + " (category: " + testCase.category + ")");
    
    // Run the test
    ValidationResult result = testRunner_->runSingleTest(testCase);
    
    // Basic assertions
    EXPECT_NE(result.status, ValidationResult::Status::ERROR) 
        << "Test should not encounter errors: " << result.message;
    
    if (result.status == ValidationResult::Status::ERROR) {
        GTEST_SKIP() << "Test encountered error: " << result.message;
    }
    
    if (result.status == ValidationResult::Status::SKIP) {
        GTEST_SKIP() << "Test was skipped: " << result.message;
    }
    
    // Main validation
    EXPECT_EQ(result.status, ValidationResult::Status::PASS) 
        << "Layout validation failed: " << result.message
        << " (Success rate: " << std::fixed << std::setprecision(1) 
        << (result.successRate() * 100) << "%)";
    
    // Additional detailed assertions for debugging
    if (result.status == ValidationResult::Status::FAIL) {
        std::cout << "\nDetailed validation results for " << testCase.name << ":" << std::endl;
        std::cout << "  Total properties: " << result.totalProperties << std::endl;
        std::cout << "  Passed: " << result.passedProperties << std::endl;
        std::cout << "  Failed: " << result.failedProperties << std::endl;
        
        // Show failed properties for debugging
        int failedCount = 0;
        for (const auto& prop : result.propertyComparisons) {
            if (!prop.withinTolerance && failedCount < 5) { // Limit output
                std::cout << "  FAILED: " << prop.property 
                         << " (expected: " << prop.expected 
                         << ", actual: " << prop.actual 
                         << ", diff: " << prop.difference 
                         << ", tolerance: " << prop.tolerance << ")" << std::endl;
                failedCount++;
            }
        }
        if (result.failedProperties > 5) {
            std::cout << "  ... and " << (result.failedProperties - 5) << " more failures" << std::endl;
        }
    }
}

// Category-specific test suites
class BasicLayoutTests : public LayoutTestSuite {};
class IntermediateLayoutTests : public LayoutTestSuite {};
class AdvancedLayoutTests : public LayoutTestSuite {};

TEST_F(BasicLayoutTests, RunAllBasicTests) {
    auto results = testRunner_->runCategory("basic");
    
    EXPECT_GT(results.totalTests, 0) << "Should have basic tests to run";
    
    // We expect at least 80% success rate for basic tests
    double successRate = static_cast<double>(results.passedTests) / results.totalTests;
    EXPECT_GE(successRate, 0.8) 
        << "Basic tests should have at least 80% success rate. "
        << "Passed: " << results.passedTests << "/" << results.totalTests
        << " (" << std::fixed << std::setprecision(1) << (successRate * 100) << "%)";
    
    // Generate detailed report
    std::string reportFile = testReportsDir_ + "/basic_tests_report.json";
    testRunner_->generateJsonReport(results, reportFile);
    
    if (::testing::Test::HasFailure()) {
        std::cout << "\nBasic test results summary:" << std::endl;
        testRunner_->printSummary(results);
    }
}

TEST_F(IntermediateLayoutTests, RunAllIntermediateTests) {
    auto results = testRunner_->runCategory("intermediate");
    
    EXPECT_GT(results.totalTests, 0) << "Should have intermediate tests to run";
    
    // We expect at least 70% success rate for intermediate tests
    double successRate = static_cast<double>(results.passedTests) / results.totalTests;
    EXPECT_GE(successRate, 0.7) 
        << "Intermediate tests should have at least 70% success rate. "
        << "Passed: " << results.passedTests << "/" << results.totalTests
        << " (" << std::fixed << std::setprecision(1) << (successRate * 100) << "%)";
    
    // Generate detailed report
    std::string reportFile = testReportsDir_ + "/intermediate_tests_report.json";
    testRunner_->generateJsonReport(results, reportFile);
    
    if (::testing::Test::HasFailure()) {
        std::cout << "\nIntermediate test results summary:" << std::endl;
        testRunner_->printSummary(results);
    }
}

TEST_F(AdvancedLayoutTests, RunAllAdvancedTests) {
    auto results = testRunner_->runCategory("advanced");
    
    EXPECT_GT(results.totalTests, 0) << "Should have advanced tests to run";
    
    // We expect at least 60% success rate for advanced tests (they're complex!)
    double successRate = static_cast<double>(results.passedTests) / results.totalTests;
    EXPECT_GE(successRate, 0.6) 
        << "Advanced tests should have at least 60% success rate. "
        << "Passed: " << results.passedTests << "/" << results.totalTests
        << " (" << std::fixed << std::setprecision(1) << (successRate * 100) << "%)";
    
    // Generate detailed report
    std::string reportFile = testReportsDir_ + "/advanced_tests_report.json";
    testRunner_->generateJsonReport(results, reportFile);
    
    if (::testing::Test::HasFailure()) {
        std::cout << "\nAdvanced test results summary:" << std::endl;
        testRunner_->printSummary(results);
    }
}

// Performance tests
class LayoutPerformanceTests : public LayoutTestSuite {};

TEST_F(LayoutPerformanceTests, TestExecutionPerformance) {
    auto startTime = std::chrono::high_resolution_clock::now();
    
    auto results = testRunner_->runCategory("basic");
    
    auto endTime = std::chrono::high_resolution_clock::now();
    double totalTime = std::chrono::duration<double>(endTime - startTime).count();
    
    // Basic tests should complete in reasonable time (10 seconds max)
    EXPECT_LT(totalTime, 10.0) << "Basic tests took too long: " << totalTime << " seconds";
    
    if (results.totalTests > 0) {
        double avgTimePerTest = totalTime / results.totalTests;
        EXPECT_LT(avgTimePerTest, 1.0) << "Average time per test too high: " << avgTimePerTest << " seconds";
    }
}

// Regression tests for specific layout features
class FlexboxLayoutTests : public LayoutTestSuite {};
class BlockLayoutTests : public LayoutTestSuite {};

// Integration tests for Radiant layout engine
class RadiantIntegrationTests : public LayoutTestSuite {};

TEST_F(FlexboxLayoutTests, ValidateFlexboxFeatures) {
    // Run tests that specifically exercise flexbox functionality
    auto allTests = testRunner_->discoverTests(".");
    
    std::vector<TestCase> flexboxTests;
    std::copy_if(allTests.begin(), allTests.end(), std::back_inserter(flexboxTests),
        [](const TestCase& test) {
            return test.name.find("flex") != std::string::npos ||
                   std::find(test.features.begin(), test.features.end(), "flexbox") != test.features.end();
        });
    
    EXPECT_GT(flexboxTests.size(), 0) << "Should have flexbox-specific tests";
    
    int passed = 0;
    int total = 0;
    
    for (const auto& test : flexboxTests) {
        auto result = testRunner_->runSingleTest(test);
        total++;
        if (result.isPass()) passed++;
        
        // Individual assertions for critical flexbox properties
        EXPECT_NE(result.status, ValidationResult::Status::ERROR) 
            << "Flexbox test should not error: " << test.name;
    }
    
    if (total > 0) {
        double successRate = static_cast<double>(passed) / total;
        EXPECT_GE(successRate, 0.75) 
            << "Flexbox tests should have at least 75% success rate";
    }
}

TEST_F(BlockLayoutTests, ValidateBlockLayoutFeatures) {
    // Run tests that specifically exercise block layout functionality
    auto allTests = testRunner_->discoverTests(".");
    
    std::vector<TestCase> blockTests;
    std::copy_if(allTests.begin(), allTests.end(), std::back_inserter(blockTests),
        [](const TestCase& test) {
            return test.name.find("block") != std::string::npos ||
                   test.name.find("margin") != std::string::npos ||
                   test.name.find("padding") != std::string::npos ||
                   std::find(test.features.begin(), test.features.end(), "block-layout") != test.features.end();
        });
    
    EXPECT_GT(blockTests.size(), 0) << "Should have block layout specific tests";
    
    int passed = 0;
    int total = 0;
    
    for (const auto& test : blockTests) {
        auto result = testRunner_->runSingleTest(test);
        total++;
        if (result.isPass()) passed++;
        
        EXPECT_NE(result.status, ValidationResult::Status::ERROR) 
            << "Block layout test should not error: " << test.name;
    }
    
    if (total > 0) {
        double successRate = static_cast<double>(passed) / total;
        EXPECT_GE(successRate, 0.8) 
            << "Block layout tests should have at least 80% success rate";
    }
}

TEST_F(RadiantIntegrationTests, BasicFlexContainerCreation) {
    // Test basic flex container creation and layout
    TestCase basicFlexTest;
    basicFlexTest.name = "basic_flex_container";
    basicFlexTest.category = "integration";
    basicFlexTest.htmlContent = R"(
        <style>
            .container {
                display: flex;
                width: 400px;
                height: 200px;
                gap: 10px;
            }
            .item {
                width: 100px;
                height: 50px;
                flex-grow: 1;
            }
        </style>
        <div class="container">
            <div class="item"></div>
            <div class="item"></div>
            <div class="item"></div>
        </div>
    )";
    
    // Test that the test case is created properly
    EXPECT_FALSE(basicFlexTest.htmlContent.empty()) << "HTML content should not be empty";
    EXPECT_EQ(basicFlexTest.name, "basic_flex_container") << "Test name should be set correctly";
    EXPECT_EQ(basicFlexTest.category, "integration") << "Test category should be set correctly";
    
    // Now test the actual Radiant integration
    testRunner_->setVerbose(true);  // Enable verbose output for debugging
    ValidationResult result = testRunner_->runSingleTest(basicFlexTest);
    EXPECT_NE(result.status, ValidationResult::Status::ERROR) 
        << "Basic flex layout should not error: " << result.message;
}

TEST_F(RadiantIntegrationTests, FlexDirectionColumn) {
    // Test flex-direction: column layout
    TestCase columnFlexTest;
    columnFlexTest.name = "flex_direction_column";
    columnFlexTest.category = "integration";
    columnFlexTest.htmlContent = R"(
        <style>
            .container {
                display: flex;
                flex-direction: column;
                width: 200px;
                height: 400px;
                gap: 5px;
            }
            .item {
                width: 100px;
                height: 80px;
            }
        </style>
        <div class="container">
            <div class="item"></div>
            <div class="item"></div>
        </div>
    )";
    
    // Test CSS parsing
    EXPECT_TRUE(columnFlexTest.htmlContent.find("flex-direction: column") != std::string::npos) 
        << "Should contain column flex direction";
    
    // TODO: Uncomment when Radiant integration is fully working
    // ValidationResult result = testRunner_->runSingleTest(columnFlexTest);
    // EXPECT_NE(result.status, ValidationResult::Status::ERROR) 
    //     << "Column flex layout should not error: " << result.message;
}

TEST_F(RadiantIntegrationTests, JustifyContentCenter) {
    // Test justify-content: center alignment
    TestCase centerJustifyTest;
    centerJustifyTest.name = "justify_content_center";
    centerJustifyTest.category = "integration";
    centerJustifyTest.htmlContent = R"(
        <style>
            .container {
                display: flex;
                justify-content: center;
                width: 300px;
                height: 100px;
            }
            .item {
                width: 50px;
                height: 50px;
            }
        </style>
        <div class="container">
            <div class="item"></div>
            <div class="item"></div>
        </div>
    )";
    
    // Test CSS parsing
    EXPECT_TRUE(centerJustifyTest.htmlContent.find("justify-content: center") != std::string::npos) 
        << "Should contain center justify content";
    
    // TODO: Uncomment when Radiant integration is fully working
    // ValidationResult result = testRunner_->runSingleTest(centerJustifyTest);
    // EXPECT_NE(result.status, ValidationResult::Status::ERROR) 
    //     << "Center justify layout should not error: " << result.message;
}

// Test data generation for parametric tests
std::vector<TestCase> generateTestParameters() {
    TestRunner runner;
    return runner.discoverTests(".");
}

// Instantiate parametric tests for all discovered test cases
INSTANTIATE_TEST_SUITE_P(
    AllLayoutTests,
    LayoutTestCase,
    ::testing::ValuesIn(generateTestParameters()),
    [](const ::testing::TestParamInfo<TestCase>& info) {
        // Generate test name from test case
        std::string name = info.param.category + "_" + info.param.name;
        
        // Replace invalid characters for GTest naming
        std::replace_if(name.begin(), name.end(), 
                       [](char c) { return !std::isalnum(c); }, '_');
        
        return name;
    }
);

// Main function for standalone execution
int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    
    // Print test suite information
    std::cout << "Radiant Layout Engine Test Suite" << std::endl;
    std::cout << "=================================" << std::endl;
    std::cout << "Running automated layout validation against browser references" << std::endl;
    std::cout << std::endl;
    
    // Change to test directory if specified
    if (argc > 1) {
        std::string testDir = argv[1];
        if (std::filesystem::exists(testDir)) {
            std::filesystem::current_path(testDir);
            std::cout << "Changed to test directory: " << testDir << std::endl;
        }
    }
    
    // Check test environment
    if (!std::filesystem::exists("./data") || !std::filesystem::exists("./reference")) {
        std::cerr << "Error: Test data directories not found." << std::endl;
        std::cerr << "Expected: ./data and ./reference directories" << std::endl;
        std::cerr << "Please run from the test/layout directory or specify correct path." << std::endl;
        return 1;
    }
    
    return RUN_ALL_TESTS();
}
